metaflow/metaflow_config.py

```diff
@@ -7,363 +6,278 @@
-        value = value_from_env
-    else:
-        value = value_from_config
-    if validate_fn and value is not None:
-        validate_fn(name, value)
-    return value
-
-
-def _get_validate_choice_fn(choices):
-    """Returns a validate_fn for use with from_conf().
-    The validate_fn will check a value against a list of allowed choices.
-    """
-
-    def _validate_choice(name, value):
-        if value not in choices:
-            raise MetaflowException(
-                "%s must be set to one of %s. Got '%s'." % (name, choices, value)
-            )
-
-    return _validate_choice
-
-
 ###
 # Default configuration
 ###
-DEFAULT_DATASTORE = from_conf("METAFLOW_DEFAULT_DATASTORE", "local")
-DEFAULT_ENVIRONMENT = from_conf("METAFLOW_DEFAULT_ENVIRONMENT", "local")
-DEFAULT_EVENT_LOGGER = from_conf("METAFLOW_DEFAULT_EVENT_LOGGER", "nullSidecarLogger")
-DEFAULT_METADATA = from_conf("METAFLOW_DEFAULT_METADATA", "local")
-DEFAULT_MONITOR = from_conf("METAFLOW_DEFAULT_MONITOR", "nullSidecarMonitor")
-DEFAULT_PACKAGE_SUFFIXES = from_conf("METAFLOW_DEFAULT_PACKAGE_SUFFIXES", ".py,.R,.RDS")
-DEFAULT_AWS_CLIENT_PROVIDER = from_conf("METAFLOW_DEFAULT_AWS_CLIENT_PROVIDER", "boto3")
+
+# Propagate defaults to be able to read what the user sets in configuration files/envs;
+# this will *not* propagate things set on the command line
+DEFAULT_DATASTORE = from_conf("DEFAULT_DATASTORE", "local")
+DEFAULT_ENVIRONMENT = from_conf("DEFAULT_ENVIRONMENT", "local")
+DEFAULT_EVENT_LOGGER = from_conf("DEFAULT_EVENT_LOGGER", "nullSidecarLogger")
+DEFAULT_METADATA = from_conf("DEFAULT_METADATA", "local")
+DEFAULT_MONITOR = from_conf("DEFAULT_MONITOR", "nullSidecarMonitor")
+DEFAULT_PACKAGE_SUFFIXES = from_conf("DEFAULT_PACKAGE_SUFFIXES", ".py,.R,.RDS")
+DEFAULT_AWS_CLIENT_PROVIDER = from_conf("DEFAULT_AWS_CLIENT_PROVIDER", "boto3")
 
 
 ###
 # Datastore configuration
 ###
 # Path to the local directory to store artifacts for 'local' datastore.
 DATASTORE_LOCAL_DIR = ".metaflow"
-DATASTORE_SYSROOT_LOCAL = from_conf("METAFLOW_DATASTORE_SYSROOT_LOCAL")
+DATASTORE_SYSROOT_LOCAL = from_conf("DATASTORE_SYSROOT_LOCAL")
 # S3 bucket and prefix to store artifacts for 's3' datastore.
-DATASTORE_SYSROOT_S3 = from_conf("METAFLOW_DATASTORE_SYSROOT_S3")
+DATASTORE_SYSROOT_S3 = from_conf("DATASTORE_SYSROOT_S3")
 # Azure Blob Storage container and blob prefix
-DATASTORE_SYSROOT_AZURE = from_conf("METAFLOW_DATASTORE_SYSROOT_AZURE")
+DATASTORE_SYSROOT_AZURE = from_conf("DATASTORE_SYSROOT_AZURE")
 
 
 ###
 # Datastore local cache
 ###
+# Cache is only local on the host machine so do not propagate values
+
 # Path to the client cache
-CLIENT_CACHE_PATH = from_conf("METAFLOW_CLIENT_CACHE_PATH", "/tmp/metaflow_client")
+CLIENT_CACHE_PATH = from_conf("CLIENT_CACHE_PATH", "/tmp/metaflow_client")
 # Maximum size (in bytes) of the cache
-CLIENT_CACHE_MAX_SIZE = int(from_conf("METAFLOW_CLIENT_CACHE_MAX_SIZE", 10000))
+CLIENT_CACHE_MAX_SIZE = from_conf("CLIENT_CACHE_MAX_SIZE", 10000)
```

Comment:
This value is no longer an int. rather than changing the names and types everywhere, we could have instead just emitted the env vars into a dict as a side effect from the `from_conf` method. That would have truly minimized the overall effort this PR has needed.

Reply:
Why is it no longer an int? It most definitely is still an int. The type of the value is now determined based on the type of the default. So in this case, 10000 is an int so the overall result will be an int. The env_vars are emitted in a dict of sort that is accessible using the `config_values` function. I can change it but at least for now `config_values` emits things as strings but the from_conf returns the proper type (serializing and deserializing happen automatically).

Reply:
perfect!